README.md
# alda-musician

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-12-12
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-12-12
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/alda-musician/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/alda-musician/compare/0.1.0...0.1.1


test/alda_musician/core_test.clj
(ns alda-musician.core-test
  (:require [clojure.test :refer :all]
            [alda-musician.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


_scratch/alda-twin-notations.clj
(require '[alda.core :refer :all])

(alda "up")


(alda "play" "-c" "piano: c d e f")


(play!
 (tempo 120)
 (part "piano")
 (octave! 4)
  (note (pitch :c))
  (note (pitch :d ))
  (note (pitch :e ))
  (note (pitch :f)))



;;;;;;;;


(alda "play" "-c" "piano: g a b > c")


(play!
 (tempo 120)
 (part "piano")
 (octave 4)
 (note (pitch :g))
 (note (pitch :a ))
 (note (pitch :b ))
 (octave :up)
 (note (pitch :c)))


;;;;;;;;



(alda "play" "-c" "piano: o0 c > c > c > c > c > c > c > c > c > c")

(play!
 (tempo 120)
 (part "piano")
 (octave 0)
 (note (pitch :c))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c ))
 (octave :up)
 (note (pitch :c)))





(alda "play" "-c" "piano: o4 c < b- a g f+")

(play!
 (tempo 120)
 (part "piano")
 (octave 4)
 (note (pitch :c))
 (octave :down)
 (note (pitch :b :flat ))
 (note (pitch :a ))
 (note (pitch :g ))
 (note (pitch :f :sharp)))



;;;;;

(alda "play" "-c" "piano: g+ f+ e+ d+ c++")

(play!
 (tempo 120)
 (part "piano")
 (octave 4)
 (note (pitch :g :sharp))
 (note (pitch :f :sharp))
 (note (pitch :e :sharp))
 (note (pitch :d :sharp))
 (note (pitch :c :sharp :sharp)))




(alda "play" "-c" "piano: o4 c4 c8 c c16 c c c c32 c c c c c c c | c1")


(play!
 (tempo 120)
 (part "piano")
 (octave 4)
 (note (pitch :c) (duration (note-length 4)))
 (note (pitch :c) (duration (note-length 8)))
 (note (pitch :c))
 (note (pitch :c) (duration (note-length 16)))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c) (duration (note-length 32)))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c))
 (note (pitch :c))
;; Bar lines are optional in Alda ( | )
 (note (pitch :c) (duration (note-length 1))))


;;;;;;;



(alda "play" "-c" "piano: r2 c | r4 c r8 c r4")


(play!
 (tempo 120)
 (part "piano")
 (octave 4)
 (note (pitch :r) (duration (note-length 2)))
 (note (pitch :c))
 (barline)
 (note (pitch :r) (duration (note-length 4)))
 (note (pitch :c))
 (note (pitch :r) (duration (note-length 8)))
 (note (pitch :c))
 (note (pitch :r) (duration (note-length 4))))



;;;;

(alda "play" "-c" "trombone: o2 c4.. d16 e-8 r c r")

(play!
 (tempo 120)
 (part "trombone")
 (octave 2)
; (voice 1)
; (panning 25)
 (note (pitch :c) (duration (note-length 4 {:dots 2})))
 (note (pitch :d) (duration (note-length 16)))
 (note (pitch :e :flat) (note-length 8)
 (note (pitch :r))
 (note (pitch :c))
 (note (pitch :r))))


;;;;


(alda "play" "-c" "piano: o2 g+1~1")

(play!
 (tempo 120)
 (part "piano")
 (octave 2)
 (note (pitch :g :sharp) (duration (note-length 1)) :slur))


;;;;


(alda "play" "-c" "piano: o4 c1/e-/g/b")

(play!
 (tempo! 120)
 (part "piano")
 (octave! 4)
 (chord
  (note (pitch :c)
        (duration (note-length 1)))
  (note (pitch :e :flat))
  (note (pitch :g ))
  (note (pitch :b #_:natural))))

;;;;



(alda "play" "-c" "piano: o4 c1/e/g/>c4 < b a g | < g+1/b/>e")

(play!
 (tempo! 120)
 (part "piano")
 (octave 4)
 (chord
 (note (pitch :c)
        (duration (note-length 1)))
  (note (pitch :e ))
  (note (pitch :g ))
  (octave :up))
 (note (pitch :c)
       (note-length 4))
 (octave :down)
 (note (pitch :b ))
 (note (pitch :a ))
 (note (pitch :g )))


;;;;



(alda "play" "-c" "V1: o5 c4 < b a g | e1")

(play!
 (tempo! 120)
 (part "piano")
 (voice 1)
 (octave 5)
 (note (pitch :c) (note-length 4))
 (octave :down)
 (note (pitch :b ))
 (note (pitch :a ))
 (note (pitch :g ))
 (barline)
 (note (pitch :e) (note-length 1)))



(alda "play" "-c" "V2: o4 c1/e/g | < g+/b")



(play!
 (tempo! 120)
 (part "piano")
 (voice 2)
 (octave 4)
 (chord
 (note (pitch :c) (note-length 1))
 (note (pitch :e ))
 (note (pitch :g )))
 (barline)
 (octave :down)
 (chord
  (note (pitch :g :sharp))
  (note (pitch :b))))









(alda "play" "-c" "bassoon: o2 d8 e (quant 30) f+ g (quant 99) a2")

(play!
; (tempo 120)
 (part "bassoon")
 (octave 2)
 (note (pitch :d) (note-length 8))
 (note (pitch :e ))
 (quant 30)
 (note (pitch :f :sharp))
 (note (pitch :g ))
 (quant 99)
 (note (pitch :a) (note-length 2)))



(alda "play" "-c" "bassoon: o2 d8 e (quant 30) (vol 65) f+ g (quant 99) a2")



(alda "play" "-c" "trumpet:  o4 c8 d e f g a b > c4.")
(alda "play" "-c" "trombone: o3 e8 f g a b > c d e4.")

(alda "play" "-c" "trumpet:  o4 c8 d e f g")
(alda "play" "-c" "trombone: o3 e8 f g a b")

(alda "play" "-c" "trumpet:  a b > c4.")
(alda "play" "-c" "trombone: > c d e4.")



(alda "play" "-c" "trumpet:  (tempo 200) o4 c8 d e f g a b > c4.")
(alda "play" "-c" "trombone: o3 e8 f g a b > c d e4.")
(alda "play" "-c" "trombone: o3 e8 f g a b > c d e4.")




(alda "play" "-c" "trumpet:  (tempo! 200) o4 c8 d e f g a b > c4.")
(alda "play" "-c" "trombone: o3 e8 f g a b > c d e4.")



(alda "play" "-c" "trumpet:  o4 c8 d e f g a b > %last-note c4.~2")
(alda "play" "-c" "trombone: o3 e8 f g a b > c d e4.~2")
(alda "play" "-c" "tuba: @last-note o2 c4.~2")


_scratch/scratch.clj
(require '[alda.core :refer :all])

(play!
 (part "piano")
 (for [notes [[:c :e :g] [:c :f :a] [:c :e :g]]]
   (apply chord (map #(note (pitch %)) notes))))


;; conjure a guitar
(play! (part "guitar"))

;; play a few notes on the guitar
(play!
 (note (pitch :e) (note-length 8))
 (note (pitch :f :sharp))
 (note (pitch :g)))

;; play a few notes, still on the guitar
(play!
 (note (pitch :a))
 (note (pitch :b) (note-length 2)))


;;;;;;; EXAMPLES

#!/usr/bin/env clojure
;; vim: ft=clojure

(require '[alda.core :refer :all])

(comment
  "Clapping Music (1972)
   for two performers

   Steve Reich

   sheet music:
   https://sites.ualberta.ca/~michaelf/SEM-O/SEM-O_2014/Steve's%20piece/clapping_reich.jpg")

(def pattern
  (let [clap "o2 d+8"
        rest (pause (duration (note-length 8)))]
    [clap clap clap rest clap clap rest clap rest clap clap rest]))

(defn shift [n pattern]
  (take (count pattern) (drop n (cycle pattern))))

(play!
 (tempo! 172)

 (part "midi-percussion")

 (voice 1)
 (repeat 13
         (repeat 12 pattern))

 (voice 2)
 (for [n (range 13)]
   (repeat 12 (shift n pattern))))

(println "Press Ctrl-C to stop & exit.")

;; (.addShutdownHook (Runtime/getRuntime)
;;                   (Thread. #(stop!)))

(stop!)

;;;;;;

;; #!/usr/bin/env clojure
;; vim: ft=clojure

(require '[alda.core :refer :all])

(def REST-RATE 0.15)
(def MS-LOWER 30)
(def MS-UPPER 3000)
(def MAX-OCTAVE 8)

(defn random-note
  "Returns a random note in a random octave with a random duration in
  milliseconds.

  May randomly return a rest with a random duration in milliseconds, instead."
  []
  (let [ms (ms (rand-nth (range MS-LOWER MS-UPPER)))]
    (if (< (rand) REST-RATE)
      (pause (duration ms))
      (let [o (rand-int (inc MAX-OCTAVE))
            n [(keyword (str (rand-nth "abcdefg")))
               (rand-nth [:sharp :flat :natural])]]
        [(octave o)
         (note (apply pitch n) (duration ms))]))))

(->str
(play!
  (part "midi-electric-piano-1")
  (panning 25)
  (repeatedly 50 random-note)

  ;; (part "midi-timpani")
  ;; (panning 50)
  ;; (repeatedly 50 random-note)

  ;; (part "midi-celesta")
  ;; (panning 75)
  #_(repeatedly 50 random-note)))

(println "Press Ctrl-C to stop & exit.")

(.addShutdownHook (Runtime/getRuntime)
  (Thread. #(stop!)))

;;;;;;;;;;;

;; #!/usr/bin/env clojure
;; vim: ft=clojure

(require '[alda.core :refer :all])

(play!
 (tempo! 180)

 (part "piano")

 (for [mode [:ionian :dorian :phrygian :lydian :mixolydian :aeolian :locrian]]
   [(key-sig [:c mode])
    "o4 c4 d8 e f g a b > c1"]))

(println "Press Ctrl-C to stop & exit.")

(.addShutdownHook (Runtime/getRuntime)
                  (Thread. #(stop!)))


;;;;;;

(->str
(play!
 (part "accordion")
       (note (pitch :c) (duration (note-length 8)))
       (note (pitch :d))
       (note (pitch :e :flat))
       (note (pitch :f))
       (note (pitch :g))
       (note (pitch :a :flat))
       (note (pitch :b))
       (octave :up)
       (note (pitch :c))))

;;;;;;


(play!
 (tempo! 180)

 (part "piano")
 (note (pitch :e :flat))
; (octave :down)
 (note (pitch :d)))


;;;;;

(play!
 (tempo! 120)
 (part "cello")
 (note (pitch :e )
       (duration (note-length 8)))
 (note (pitch :f ))
 (note (pitch :g )))
;;;;;;;;;;;;;;



(alda "play" "-c" "piano: c1 / e- / g / b")


(play!
(tempo! 120)
(part "piano")
(octave! 4)
(chord
 (note (pitch :c)
       (duration (note-length 1)))
 (note (pitch :e :flat))
 (note (pitch :g ))
 (note (pitch :b #_:natural))))



;;;;;;;;;;;;;;
;; Qualified import
;;;;;;;;;;;;;;

(require '[alda.core :as alda])

(println (alda/alda "version"))

(println (alda/alda "status"))


